### PR TITLE
Add custom exceptions for directories.

### PR DESCRIPTION
**Pull request by ibradwan**

### CODE REVIEW

File: src/Common/Exceptions/BaseException.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#ifndef __BASE_EXCEPTION_H_
+#define __BASE_EXCEPTION_H_
+
+#include <iostream>
```
Comment (OmarBazaraa) [inline src/Common/Exceptions/BaseException.h]:
```suggestion
#include <string>
```

File: src/Common/Exceptions/BaseException.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace FileManager {
+
+    class BaseException {
```
Comment (OmarBazaraa) [inline src/Common/Exceptions/BaseException.h]: Shouldn't this be inheriting from the system exception class?

File: src/Common/Exceptions/BaseException.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    public:
+
+        std::string message;
```
Comment (OmarBazaraa) [inline src/Common/Exceptions/BaseException.h:11]: Can we make this `private`?

File: src/Common/Exceptions/DirectoryExistingException.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#include "BaseException.h"
+
+namespace FileManager {
+
+    class DirectoryExistingException : public BaseException {
+    public:
+
+        std::string path;
+
+        DirectoryExistingException(std::string path)
```
Comment (OmarBazaraa) [inline src/Common/Exceptions/DirectoryExistingException.h]: What about?

```suggestion
        DirectoryAlreadyExistsException(std::string path)
```

**Review by OmarBazaraa — CHANGES_REQUESTED** [review]

File: src/Common/Exceptions/DirectoryExistingException.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        : BaseException("Directory " + path + " already exists.") {
```
Comment (OmarBazaraa) [inline src/Common/Exceptions/DirectoryExistingException.h]: I am thinking to change the convention a little bit when specifying the initialization list ..
What do you think?

```suggestion
        : BaseException("Directory " + path + " already exists.")
        {
```

File: src/Common/Exceptions/BaseException.h (hunk 2 above)
Comment (ibradwan) [reply src/Common/Exceptions/BaseException.h]: Yeah sorry 😀

File: src/Common/Exceptions/DirectoryExistingException.h (hunk 4 above)
Comment (ibradwan) [reply src/Common/Exceptions/DirectoryExistingException.h]: Thought about it, but I feel it's longer than needed.

File: src/Common/Exceptions/DirectoryExistingException.h (hunk 5 above)
Comment (ibradwan) [reply src/Common/Exceptions/DirectoryExistingException.h]: Mmmm I actually prefer it to always keep the `{` on a new line. But anyway, I prefer it to be consistent with(out) the initialization list.

File: src/Common/Exceptions/DirectoryExistingException.h (hunk 4 above)
Comment (OmarBazaraa) [reply src/Common/Exceptions/DirectoryExistingException.h]: Not that much 😁

File: src/Common/Exceptions/BaseException.h (hunk 3 above)
Comment (ibradwan) [reply src/Common/Exceptions/BaseException.h:11]: Mmmm we can, but I believe this should be accessible to anyone, may through a getter.

**Pull request merged**